pkg/minibroker/rabbitmq.go

```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package minibroker
+
+import (
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type RabbitmqProvider struct{}
+
+func (p RabbitmqProvider) Bind(services []corev1.Service, params map[string]interface{}, chartSecrets map[string]interface{}) (*Credentials, error) {
+	service := services[0]
+
+	var amqpPort *corev1.ServicePort = nil
+	for _, port := range service.Spec.Ports {
+		if port.Name == "amqp" {
+			amqpPort = &port
+			break
+		}
+	}
+	if amqpPort == nil {
+		return nil, errors.Errorf("no amqp port found")
+	}
+
+	var password string
+	passwordVal, ok := chartSecrets["rabbitmq-password"]
+	if !ok {
+		return nil, errors.Errorf("password not found in secret keys")
+	}
+	password = passwordVal.(string)
+
+	host := buildHostFromService(service)
+	creds := Credentials{
+		Protocol: amqpPort.Name,
+		Port:     amqpPort.Port,
+		Host:     host,
+		Username: "user",
+		Password: password,
+		Database: "%2F",
```

Comment:
It's not clear to me what `%2F` does here.

Reply:
... I see now. It's the URL encoding for a forward slash. This gets used by https://github.com/kubernetes-sigs/minibroker/blob/e50e4bc41f03ae2e10daaf9522f2aafb32bc7ea3/pkg/minibroker/provider.go#L65

The problem is that then it gets a double forward slash. According to the RabbitMQ URI spec, it doesn't require to end with a forward slash. Doesn't it work with a blank `Database`?

Reply:
The default rabbitmq vhost is `/` and the `%2F` is the URL encoded representation of that. It doesn't really become a double forward slash, it's just a "proper" slash splitting the path segments and an encoded slash as the path segment itself. There's nothing wrong with the URL itself (e.g. `amqp://user:nTC1Ff8DsY@rabbitmq-11305fe2b7b82316c52a.minibroker-tests.svc:5672/%2F`), it's just how it looks for the default vhost.

Having said that, I think I'll rather set the `Database` field to `/` (because that's how the default vhost is called after all) and handle the encoding when building the host, that makes it a lot clearer. It would also work with omitting the vhost from the URL entirely and relying on the default being used but I would prefer to have the vhost in there explicitly tbh.

Reply:
I like explicit code. Even though, it was not immediately obvious that `%2DF` would end up being in the URI since it's setting `Database`.

I worry that the clients won't understand the URL-encoded database if they decide not to use the `URI` field and create a client in their preferred language with the individual fields.

Would an explicit `/` for `Database` solve this? You can always construct the URI yourself. You don't need to use the function, so you can then use `%2F` in the URI explicitly and still be explicit in the `Database`.

Reply:
Oh... you force-pushed the commits. I was looking at old code. :|
You already did what I just described. LOL